pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,479 @@
+
+    String hadoopTokenFileLocation = System.getenv("HADOOP_TOKEN_FILE_LOCATION");
+    if (hadoopTokenFileLocation != null) {
+      job.getConfiguration().set("mapreduce.job.credentials.binary", hadoopTokenFileLocation);
+    }
+
+    // Schema configs.
+    Schema schema = getSchema(inputDataPath.get(0));
+    _logger.info("Schema is: {}", schema.toString(true));
+
+    // Mapper configs.
+    job.setMapperClass(SegmentPreprocessingMapper.class);
+    job.setMapOutputKeyClass(AvroKey.class);
+    job.setMapOutputValueClass(AvroValue.class);
+    job.getConfiguration().setInt(JobContext.NUM_MAPS, inputDataPath.size());
+
+    // Reducer configs.
+    job.setReducerClass(SegmentPreprocessingReducer.class);
+    job.setOutputKeyClass(AvroKey.class);
+    job.setOutputValueClass(NullWritable.class);
+    if (_properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS) != null) {
+      _logger.info("Setting {} to {}", MAXIMUM_NUMBER_OF_RECORDS, _properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS));
+      job.getConfiguration().set(MAXIMUM_NUMBER_OF_RECORDS, _properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS));
+    }
+    AvroMultipleOutputs.addNamedOutput(job, "avro", AvroKeyOutputFormat.class, schema);
+    AvroMultipleOutputs.setCountersEnabled(job, true);
+    LazyOutputFormat.setOutputFormatClass(job, AvroKeyOutputFormat.class);
+
+    // Input and output paths.
+    FileInputFormat.setInputPaths(job, _inputSegmentDir);
+    FileOutputFormat.setOutputPath(job, outputDir);
+
+    List<FileStatus> inputDataFiles = new ArrayList<>();
+    for (String inputPath : inputDataPath) {
+      Path inputPathPattern = new Path(inputPath);
+      FileStatus[] fileStatusArr = _fileSystem.globStatus(inputPathPattern);
+      for (FileStatus fileStatus : fileStatusArr) {
+        inputDataFiles.addAll(getDataFilesFromPath(_fileSystem, fileStatus.getPath()));
+      }
+    }
+    _logger.info("Total number of files to be pre-computed: {}", inputDataFiles.size());
+
+    // Set up mapper output key
+    Set<Schema.Field> fieldSet = new HashSet<>();
+
+    // Partition configs.
+    int numReduceTasks = (_numberOfPartitions != 0) ? _numberOfPartitions : inputDataFiles.size();
+    if (_partitionColumn != null) {
+      job.getConfiguration().set(ENABLE_PARTITIONING, "true");
+      job.setPartitionerClass(GenericPartitioner.class);
+      job.getConfiguration().set(PARTITION_COLUMN, _partitionColumn);
+      if (_partitionFunction != null) {
+        job.getConfiguration().set(PARTITION_FUNCTION, _partitionFunction);
+      }
+      job.getConfiguration().set(NUMBER_OF_PARTITIONS, Integer.toString(numReduceTasks));
+    } else {
+      if (_numberOfOutputFiles > 0) {
+        numReduceTasks = _numberOfOutputFiles;
+      }
+      // Partitioning is disabled. Adding hashcode as one of the fields to mapper output key.
+      // so that all the rows can be spread evenly.
+      addHashCodeField(fieldSet);
+    }
+    job.setInputFormatClass(CombineAvroKeyInputFormat.class);
```

Comment:
can it happen that files are repartitioned once more when sorted or do you always do 1:1 file mapping after partitioning? Seems not right, because we always use the comnbine avro key format?

also, does min # output files/max records change how files are partitioned? could that break partitioning?

Reply:
The config `min # output files` will be **disabled** if partitioning is **enabled**.